android/java/org/chromium/chrome/browser/settings/BraveAddNewsSources.java

```diff
@@ -0,0 +1,159 @@
+
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.settings;
+
+import android.content.SharedPreferences;
+import android.os.Bundle;
+import android.view.KeyEvent;
+import android.view.inputmethod.EditorInfo;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.preference.CheckBoxPreference;
+import androidx.preference.EditTextPreference;
+import androidx.preference.Preference;
+import androidx.preference.PreferenceCategory;
+import androidx.preference.PreferenceFragmentCompat;
+import androidx.preference.PreferenceGroup;
+import androidx.preference.PreferenceManager;
+import androidx.preference.PreferenceScreen;
+import androidx.preference.SwitchPreference;
+
+import org.chromium.base.ContextUtils;
+import org.chromium.base.Log;
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.BraveLaunchIntentDispatcher;
+import org.chromium.chrome.browser.app.BraveActivity;
+import org.chromium.chrome.browser.preferences.BravePreferenceKeys;
+import org.chromium.chrome.browser.settings.BravePreferenceFragment;
+import org.chromium.components.browser_ui.settings.ChromeBaseCheckBoxPreference;
+import org.chromium.components.browser_ui.settings.ChromeSwitchPreference;
+import org.chromium.components.browser_ui.settings.SettingsUtils;
+
+import org.chromium.brave_news.mojom.Publisher;
+import org.chromium.brave_news.mojom.BraveNewsController;
+import org.chromium.chrome.browser.brave_news.BraveNewsControllerFactory;
+import org.chromium.mojo.bindings.ConnectionErrorHandler;
+import org.chromium.mojo.system.MojoException;
+import java.util.Map;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class BraveAddNewsSources
+        extends BravePreferenceFragment implements Preference.OnPreferenceChangeListener,
+        ConnectionErrorHandler {
+    private static final String PREF_ADD_SOURCES = "news_source_1";
+    private EditTextPreference addSource;
+    private EditText mEditText;
+    private PreferenceScreen mainScreen;
+    private BraveNewsController mBraveNewsController; 
+    private ArrayList<Publisher> mPublishers;
+
+    public static int getPreferenceSummary() {
+        return BraveLaunchIntentDispatcher.useCustomTabs() ? R.string.text_on : R.string.text_off;
+    }
+
+    @Override
+    public void onConnectionError(MojoException e) {
+        mBraveNewsController = null;
+        InitBraveNewsController();
+    }
+
+    private void InitBraveNewsController() {
+        if (mBraveNewsController != null) {
+            return;
+        }
+
+        mBraveNewsController =
```

Comment:
we need to call `mBraveNewsController.close()` in onDestroy to prevent leaks in mojo

Reply:
done